### PR TITLE
Make order ticket available from order event

### LINKED ISSUE #6736: Unable to Get Order Id with Async Market Orders
It would be nice if async market orders return the order id before triggering an order event. 

#### Expected Behavior
Market orders would take 10ms to fill. 

#### Actual Behavior
Market orders taking 0ms to fill

#### Potential Solution
 - 

#### Reproducing the Problem
#region imports
from AlgorithmImports import *
#endregion
import copy

class TrailingStopLoss(QCAlgorithm):
    
    def Initialize(self):
        self.SetStartDate(2022, 11, 1)
        self.SetEndDate(2022, 11, 1)
        self.SetCash(100000)
        self.equity = self.AddEquity("QQQ", Resolution.Second)
        self.symbol = self.equity.Symbol
        self.Consolidate(self.symbol, timedelta(minutes=1), self.minConsolidator)

        # ignore fees for now
        self.equity.SetFeeModel(ConstantFeeModel(0))

        #self.Schedule.On(self.DateRules.EveryDay(self.symbol), 
        #    self.TimeRules.Every(timedelta(seconds=1)),
        #    self.closePosition)

        self.trade_count = 0
        self.Count = 0
        self.con_count = 0
        self.last_order = 'None'
        self.ticket = None

    # NOTE: consolidator event is called BEFORE the onData in case they both need to
    # be called
    # a TraderBar object is passed in
    def minConsolidator(self, bar):
        if self.trade_count < 3:
            # buy
            self.ticket = self.MarketOrder(self.symbol, 100, True)
            self.Debug(f"{self.Time}: BUY: {bar.Price}, orderId: {self.ticket.OrderId}")
            self.last_order = 'Buy'
            self.trade_count += 1

    def OnData(self, data):
        return

        if self.symbol not in data:
            return

        if not data.Bars.ContainsKey(self.symbol):
            return

        # close a previous order if there is one
        if self.Portfolio[self.symbol].Invested:
            if self.Time.second > 58:
                self.Debug(f"{self.Time}: close QQQ,{self.Securities[self.symbol].BidPrice}, {self.Securities[self.symbol].AskPrice}")
                self.Liquidate(self.symbol)
        return

    def OnOrderEvent(self, orderEvent):
        if self.ticket is not None:
            self.Debug(f"{self.Time}: OnOrder: ticket.orderId: {self.ticket.OrderId}, event.orderId: {orderEvent.OrderId}, status: {orderEvent.Status}")
        else:
            self.Debug(f"{self.Time}: OnOrder: self.ticket is None, event.orderId: {orderEvent.OrderId}, status: {orderEvent.Status}")

LOGS
----

2022-11-01 09:31:00: OnOrder: self.ticket is None, event.orderId: 1, status: 1
2022-11-01 09:31:00: OnOrder: self.ticket is None, event.orderId: 1, status: 3
2022-11-01 09:31:00: BUY: 281.69, orderId: 1

2022-11-01 09:32:00: OnOrder: ticket.orderId: 1, event.orderId: 2, status: 1
2022-11-01 09:32:00: OnOrder: ticket.orderId: 1, event.orderId: 2, status: 3
2022-11-01 09:32:00: BUY: 281.28, orderId: 2

2022-11-01 09:33:00: OnOrder: ticket.orderId: 2, event.orderId: 3, status: 1
2022-11-01 09:33:00: OnOrder: ticket.orderId: 2, event.orderId: 3, status: 3
2022-11-01 09:33:00: BUY: 281.6, orderId: 3
 
#### Checklist 
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by jhonabreul**
#### Description

Making order ticket accessible from order events.

#### Related Issue

Closes #6736 

#### How Has This Been Tested?

Regression algorithms

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Engine/TransactionHandlers/BrokerageTransactionHandler.cs [hunk 1]
```diff
@@ -1187,6 +1186,7 @@ private void HandleOrderEvents(List<OrderEvent> orderEvents)
 
                     // update the ticket after we've processed the fill, but before the event, this way everything is ready for user code
                     tickets[i].AddOrderEvent(orderEvent);
+                    orderEvent.Ticket = tickets[i];
```
Comment (Martin-Molinero) [inline Engine/TransactionHandlers/BrokerageTransactionHandler.cs:1189]: I think we might be able to get rid of the `tickets` collection entirely here, maybe we just set the ticker on the order even on the first loop at the top?

File: Algorithm.CSharp/OrderTicketAssignmentDemoAlgorithm.cs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                Debug($"{Time}: Buy: Price {bar.Price}, orderId: {_ticket.OrderId}");
+                _tradeCount++;
+            });
+        }
+
+        public override void OnOrderEvent(OrderEvent orderEvent)
+        {
+            // We cannot access _ticket directly because it is assigned asynchronously:
+            // this order event could be triggered before _ticket is assigned.
+            var ticket = orderEvent.Ticket;
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/OrderTicketAssignmentDemoAlgorithm.cs:57]: Nice! Think we could assert that `_ticket` is null here 👍

**Review by Martin-Molinero — APPROVED** [review]
Nice! Minor requests

File: Engine/TransactionHandlers/BrokerageTransactionHandler.cs (hunk 1 above)
Comment (jhonabreul) [reply Engine/TransactionHandlers/BrokerageTransactionHandler.cs:1189]: Also, when we worked on HandleOrderEvents, we needed to make sure to find every order and ticket before starting to update the orders, that being the reason of having separate loops and the collections.

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**